testng-core/src/test/java/test/listeners/ListenersTest.java

```diff
@@ -0,0 +1,139 @@
+package test.listeners;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import test.SimpleBaseTest;
+import test.listeners.issue2638.DummyInvokedMethodListener;
+import test.listeners.issue2638.TestClassASample;
+import test.listeners.issue2638.TestClassBSample;
+
+public class ListenersTest extends SimpleBaseTest {
+
+  private static final String[] github2638ExpectedList =
+      new String[] {
+        "test.listeners.issue2638.TestClassASample.testMethod",
+        "test.listeners.issue2638.TestClassBSample.testMethod"
+      };
+
+  @Test(description = "GITHUB-2638", dataProvider = "suiteProvider")
+  public void ensureDuplicateListenersAreNotWiredInAcrossSuites(
+      XmlSuite xmlSuite, Map<String, String[]> expected) {
+    TestNG testng = create(xmlSuite);
+    testng.run();
+
+    assertThat(DummyInvokedMethodListener.getMethods("Container_Suite"))
+        .containsExactly(expected.get("Container_Suite"));
+    assertThat(DummyInvokedMethodListener.getMethods("Inner_Suite1"))
+        .containsExactly(expected.get("Inner_Suite1"));
+    assertThat(DummyInvokedMethodListener.getMethods("Inner_Suite2"))
+        .containsExactly(expected.get("Inner_Suite2"));
+    DummyInvokedMethodListener.reset();
+  }
+
+  @DataProvider(name = "suiteProvider")
+  public Object[][] getSuites() throws IOException {
+    return new Object[][] {
+      new Object[] {getNestedSuitesViaXmlFiles(), getExpectations()},
+      new Object[] {getNestedSuitesViaApis(), getExpectations()},
+      new Object[] {getNestedSuitesViaXmlFilesWithListenerInChildSuite(), getExpectations()},
```

Comment:
It confirms that the current implementation of the "child suite" feature is just a merge of all suites into a single one.

Maybe we should add it to the document because I didn't find anything that explains the behavior.

Reply:
Could you confirm that only one before/after suite is triggered? (I don't know if a test already covers this case)

Reply:
@juherr 

> Could you confirm that only one before/after suite is triggered? (I don't know if a test already covers this case)

Where are you looking at this confirmation? I dont think this expectation is true/valid. It would need to be executed for every suite.

Reply:
> Where are you looking at this confirmation? I dont think this expectation is true/valid. It would need to be executed for every suite.

My question is about the actual behavior of (suite) listeners on suites (parent and/or children mixed).
In your sample, what is it? 1, 2, or 3 "before/after suite"?
Is it already covered by a test? If not, could you add one?

> Not true. The suites are not merged into it. I think the listeners are hoisted and then used as a big grand collection.

Could you elaborate? I supposed we copy the listener node from children to parent and the same for the tests node.

Reply:
@juherr 

> My question is about the actual behavior of (suite) listeners on suites (parent and/or children mixed).
In your sample, what is it? 1, 2, or 3 "before/after suite"?

Its not clear what you are trying to find out.

TestNG defines a global map for suites. See [here](https://github.com/cbeust/testng/blob/master/testng-core/src/main/java/org/testng/TestNG.java#L147).

We extract out the listeners from each suite and its child suites and keep adding them to this global map which is what again gets passed over to the SuiteRunner.

This PR is merely trying to avoid re-adding the same listener again (which is anyways getting blocked by the edit check [here](https://github.com/cbeust/testng/blob/master/testng-core/src/main/java/org/testng/TestNG.java#L683)

> Is it already covered by a test?

The existing tests that deal with listeners in child suites should cover this. I dont know if there is a test. But the tests that I have added should suffice and I dont think we need to add a suite level listener to test this out.

> Could you elaborate? I supposed we copy the listener node from children to parent and the same for the tests node.

This method should elaborate what I am hinting at https://github.com/cbeust/testng/blob/a5c29abe0981145b13c95385961bcdf8b937a899/testng-core/src/main/java/org/testng/TestNG.java#L963-L986

We extract out the listeners from suites and their child suites and then store them in a global variable within the TestNG object